File: results/0.0/ICMP/icmp-gpt-4-turbo.c
#include <hammer/hammer.h>
#include <stdio.h>
#include <stdlib.h>

// Define parsers for ICMP fields
static HParser *icmp_type;
static HParser *icmp_code;
static HParser *icmp_checksum;
static HParser *icmp_rest_of_header;
static HParser *icmp_message;

void init_parsers() {
    icmp_type = h_uint8();
    icmp_code = h_uint8();
    icmp_checksum = h_uint16();
    icmp_rest_of_header = h_bits(32, false); // Generic 32 bits for simplicity
    icmp_message = h_sequence(icmp_type, icmp_code, icmp_checksum, icmp_rest_of_header, NULL);
}

void parse_icmp(const uint8_t *data, size_t length) {
    HParseResult *result = h_parse(icmp_message, data, length);
    if (result) {
        printf("ICMP Packet Parsed Successfully:\n");
        printf("Type: %d\n", *(uint8_t*)h_value_uint(result->ast->children[0]));
        printf("Code: %d\n", *(uint8_t*)h_value_uint(result->ast->children[1]));
        printf("Checksum: %d\n", *(uint16_t*)h_value_uint(result->ast->children[2]));
        // Rest of Header is not interpreted further here
    } else {
        printf("Failed to parse ICMP packet.\n");
    }
    h_parse_result_free(result);
}

int main(int argc, char **argv) {
    if (argc != 2) {
        fprintf(stderr, "Usage: %s <icmp_binary_file>\n", argv[0]);
        return 1;
    }

    const char *filename = argv[1];
    FILE *file = fopen(filename, "rb");
    if (!file) {
        perror("Failed to open file");
        return 1;
    }

    // Seek to the end of the file to determine the file size
    fseek(file, 0, SEEK_END);
    long file_size = ftell(file);
    rewind(file);

    // Read the entire file into memory
    uint8_t *data = malloc(file_size);
    if (!data) {
        perror("Failed to allocate memory");
        fclose(file);
        return 1;
    }

    size_t read_size = fread(data, 1, file_size, file);
    if (read_size != file_size) {
        fprintf(stderr, "Error reading file\n");
        free(data);
        fclose(file);
        return 1;
    }

    // Initialize parsers
    init_parsers();

    // Parse the ICMP packet
    parse_icmp(data, file_size);

    // Cleanup
    free(data);
    fclose(file);

    return 0;
}

File: results/0.0/ICMP/icmp-gpt-4-turbo.rs
use nom::{
    IResult,
    bytes::complete::{take},
    number::complete::{be_u8, be_u16},
};
use std::env;
use std::fs::File;
use std::io::{self, Read};

#[derive(Debug)]
struct IcmpPacket {
    icmp_type: u8,
    code: u8,
    checksum: u16,
    rest_of_header: Vec<u8>,
    data: Vec<u8>,
}

fn parse_icmp_packet(input: &[u8]) -> IResult<&[u8], IcmpPacket> {
    let (input, icmp_type) = be_u8(input)?;
    let (input, code) = be_u8(input)?;
    let (input, checksum) = be_u16(input)?;
    let (input, rest_of_header) = take(4usize)(input)?;
    let (input, data) = take(input.len())(input)?;

    Ok((input, IcmpPacket {
        icmp_type,
        code,
        checksum,
        rest_of_header: rest_of_header.to_vec(),
        data: data.to_vec(),
    }))
}

fn main() -> io::Result<()> {
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        eprintln!("Usage: {} <file>", args[0]);
        std::process::exit(1);
    }

    let mut file = File::open(&args[1])?;
    let mut buffer = Vec::new();
    file.read_to_end(&mut buffer)?;

    match parse_icmp_packet(&buffer) {
        Ok((_, packet)) => {
            println!("{:?}", packet);
        },
        Err(e) => {
            eprintln!("Failed to parse ICMP packet: {:?}", e);
        }
    }

    Ok(())
}